### ISSUE TITLE
Speed

### ISSUE THREAD

**Issue by kpu** [bitextor/bicleaner#8]
I suspect bicleaner can be much much faster.  Have you done some profiling?  

As an example, you're currently running a regex for each character class to count them:
https://github.com/bitextor/bicleaner/blob/master/bicleaner/features.py#L243
which would be faster just iterating over characters, classifying, and incrementing the appropriate count.  

Are you doing any batching for classifier inference?

**Comment by sortiz** [comment]
Yes, we profiled bicleaner (the profiling code for cProfile is currently commented). The bottleneck was and is qmax, it was improved to perform a "approximated Qmax" that works the same with short sentences and OK with long sentences.

I don't remember if it was relevant or not the time that character_class_dist took. Probably it wasn't. We'll check.

Classifier inference was batched by "block" in the previous version, not in the "lite" one because it happens to be parallel.

**Comment by kpu** [comment]
As in the classifier uses multiple threads when batching?  Batching is still a good strategy even in the lite version with a single thread.  As Hieu mentioned, the current parallelization often leaves CPUs idle.

**Comment by mbanon** [comment]
Hey,
we've been doing some performance tests:

* Full bicleaner with -p1 option: 1743 rows/sec
* Lite bicleaner:  ???
* Lite bicleaner, allowing parallelization in sklearn: 3590 rows/sec
* Lite bicleaner with character_class_dist refactor : ???
* Lite bicleaner with parallelization in sklearn and character_class_dist refactor : 3541 rows/sec

(Please note that the non-parallelized lite versions are significantly slower than the ones allowing parallelization in sklearn, probably 30x slower or so)


I'd go with the lite version that allows parallelization in sklearn. If you are ok with this, I'll update the current bicleaner-classify-lite with the faster version.

**Comment by mbanon** [comment]
Both things

**Comment by kpu** [comment]
So. . . is it possible to batch computation to take advantage of SIMD but not thread (which we would handle with multiple processes)?

**Comment by mbanon** [comment]
Apparently, it's possible.
I've changed the code in the lite classifier to force sklearn to use only one job.
But, according to their own documentation, it's still possible that some paralelism occur due  to numerical processing libraries. 
In order to avoid that, please check 
https://scikit-learn.org/stable/faq.html#faq-mkl-threading

**Comment by kpu** [comment]
I'm ok with it being faster ;-)

**Comment by mbanon** [comment]
Updated bicleaner lite both in repository and PyPI  (version 0.10.6)
Closing  issue.
